Add support for multiple focus layers for Controls

redsett opened the pull request:
Added support for multiple Controls to be focused at once, via "focus layers".

Summary:
Each Control is assigned to a (int)**_focus_layer_**, and the Viewport now has a (int)**_active_focus_layer_**. A Viewport will only navigate to Controls that match the **_active_focus_layer_**. E.g. If the **_active_focus_layer_** is 1, the Viewport will only navigate to Controls with their **_focus_layer_** set to 1. You could then **_set_active_focus_layer()_** to switch to navigating Controls in another layer.


Example:
<img width="956" alt="image" src="https://user-images.githubusercontent.com/17795893/175797100-0d8af0d5-dac8-47b4-ac2e-9bc37e3d6905.png">
<img width="311" alt="image" src="https://user-images.githubusercontent.com/17795893/175797110-49fdb39f-bf4e-41cd-93af-71d36de62afe.png">

@Faless @groud
This mostly solves the split screen issue as well, but doesn't have the multiple input support by default like @Faless's does linked below, but there's no reason his solution couldn't be added on top of this.
https://github.com/godotengine/godot/pull/20091



Note: This is my first time contributing to a github project in general. So please excuse my noobness. I'm happy to learn from you all, so feel free to slap me on the wrist if need be. :)

Calinou commented:
Nice work! I believe this addresses https://github.com/godotengine/godot-proposals/issues/385 and https://github.com/godotengine/godot-proposals/issues/4295.

redsett commented:
1)
https://github.com/godotengine/godot-proposals/issues/385 this is exactly what I was adding support for on my personal project.

2)
https://github.com/godotengine/godot-proposals/issues/4295 I don't believe this fully adds support for focusing with multiple inputs. But with a bit of extra gdscript hackery it could.

Example of how you could do this while still using multiple focus layers in each Viewport:
Viewport1 only use **_active_focus_layer's_** 1-4, Viewport2 only use **_active_focus_layer_**'s 5-8. Then make sure the Controls in each Viewport is set correctly.

To fully support multiple inputs without getting fancy(hacky), we'd have to add something like this on top of it:
https://github.com/godotengine/godot/pull/20091

redsett commented:
control.cpp got a big revamp which has many conflicts with this change. I'm just going to close this request. Probably a bad time to be adding features like this one.

YuriSizov commented:
> control.cpp got a big revamp which has many conflicts with this change. I'm just going to close this request. Probably a bad time to be adding features like this one.

Sorry for the inconvenience. It should be good to make PRs now though, and the change was only superficial. You would likely need to port your work by hand, but it shouldn't require any code changes. That big PR just rearranged things.

redsett commented:
Sadly, I can't make that deadline. :/

Thanks YuriSizov, no worries.

KoBeWi commented:
Technically this was opened before the freeze and a rebased PR counts as the same feature, so...

YuriSizov commented:
Yeah, sorry, I didn't imply that you need to resubmit it before August 3rd. Just wanted to dispute that it's not the best time to contribute something like that, when in fact it's the perfect time.

redsett commented:
Understood! I'll take a look this weekend at merging.

redsett commented:
Merged

YuriSizov commented:
Can you please squash your commits, per our [guidelines](https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase)

redsett commented:
> Can you please squash your commits, per our [guidelines](https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase)

I think I did the thing.

YeldhamDev commented:
The current commit message is currently "parent 836fe9a", it should be the actual message instead.

YuriSizov commented:
And CI seems to be failing because of a couple of code style issues, and because you need to run the doctool to update some documentation properly.

KoBeWi commented:
I tested the feature. It's great, but feels like there should be an easier way to switch between layers. Maybe an optional argument to `grab_focus()` that makes it automatically switch layer when `true`?

KoBeWi commented on doc/classes/Control.xml:
```diff
@@ -549,7 +549,7 @@
 		</method>
 		<method name="grab_focus">
-			<return type="void" />
+			<return type="bool" />
```
The new return value needs to be documented (i.e. what does it do).

KoBeWi commented on doc/classes/Control.xml:
```diff
@@ -973,8 +973,12 @@
+		</member>
 		<member name="focus_mode" type="int" setter="set_focus_mode" getter="get_focus_mode" enum="Control.FocusMode" default="0">
-			The focus access mode for the control (None, Click or All). Only one Control can be focused at the same time, and it will receive keyboard signals.
+			The focus access mode for the control (None, Click or All). Only one [Control] per layer can be focused at the same time, and it will receive keyboard signals.
```
```suggestion
			The focus access mode for the control (None, Click or All). Only one [Control] per focus layer can be focused at the same time, and it will receive keyboard signals.
```

KoBeWi commented on doc/classes/Viewport.xml:
```diff
@@ -95,14 +95,20 @@
-		<method name="gui_get_focus_owner">
+		<method name="gui_get_focus_owner" qualifiers="const">
 			<return type="Control" />
+			<argument index="0" name="arg0" type="int" />
```
New argument needs to be mentioned in the description.
Also you should specify the name in `bind_method`.

KoBeWi commented on doc/classes/Control.xml:
```diff
@@ -973,8 +973,12 @@
 		</member>
+		<member name="focus_layer" type="int" setter="set_focus_layer" getter="get_focus_layer" default="0">
+			The focus layer for this control. Each layer will only navigate to other [Control]s in the same layer. Each layer can only have one focus at a time.
+			Use [code]gui_set_active_focus_layer[/code] to set the active focus layer that will be navigated.
```
```suggestion
			Use [method Viewport.gui_set_active_focus_layer] to set the active focus layer that will be navigated.
```

KoBeWi commented on scene/gui/control.cpp:
```diff
@@ -1791,19 +1791,28 @@ Control::FocusMode Control::get_focus_mode() const {
 }
 
+void Control::set_focus_layer(int layer_id) {
+	data.focus_layer = layer_id;
```
When you change layer of a focused control, it may cause multiple controls on the same layer to be focused. You need to add a check here.

KoBeWi commented on scene/main/viewport.cpp:
```diff
@@ -2183,23 +2196,60 @@ Window *Viewport::get_base_window() const {
 
-void Viewport::_gui_control_grab_focus(Control *p_control) {
-	if (gui.key_focus && gui.key_focus == p_control) {
+void Viewport::_gui_control_grab_focus(Control *p_control, int focus_layer) {
```
The second parameter here is not necessary, because you can just get the layer from the provided control.

redsett commented:
Thank you for all the feedback KoBeWi!

Most of the feedback has been addressed locally. I just need to look more into your set_focus_layer() concern.

redsett commented:
Applied the feedback.

KoBeWi commented on scene/gui/control.cpp:
```diff
@@ -1845,19 +1845,46 @@ Control::FocusMode Control::get_focus_mode() const {
+
+		if (next) {
+			next->grab_focus(false);
+			get_viewport()->set_input_as_handled();
```
What is this for?

redsett replied:
Great question. It was probably a copy/paste thing from... somewhere. Either way, I removed it and tested, works fine without it.

KoBeWi commented on doc/classes/Control.xml:
```diff
@@ -550,9 +550,11 @@
+			<param index="0" name="p_auto_set_viewport_active_focus_layer" type="bool" />
 			<description>
 				Steal the focus from another control and become the focused control (see [member focus_mode]).
+				Returns [code]true[/code] if it was able to grab focus. Would return [code]false[/code] if the control isn't set to be focused (see [member focus_mode]).
```
Doesn't it also return `false` if the controls is on a wrong layer? It should be mentioned too.

KoBeWi commented on scene/gui/control.cpp:
```diff
@@ -1845,19 +1845,46 @@ Control::FocusMode Control::get_focus_mode() const {
+
+		Control *next = find_next_valid_focus();
+
+		if (next) {
```
```suggestion
		if (next && next != this) {
```
Otherwise the control doesn't lose focus properly.

KoBeWi commented on doc/classes/Viewport.xml:
```diff
@@ -122,10 +128,17 @@
 				Removes the focus from the currently focused [Control] within this viewport. If no [Control] has the focus, does nothing.
 			</description>
 		</method>
+		<method name="gui_set_active_focus_layer">
+			<return type="void" />
+			<param index="0" name="p_focus_layer" type="int" />
+			<description>
+			</description>
+		</method>
```
Missing description; would be nice to add it.

KoBeWi commented on doc/classes/Viewport.xml:
```diff
@@ -95,14 +95,20 @@
 			<return type="Control" />
+			<param index="0" name="p_focus_layer" type="int" />
 			<description>
 				Returns the [Control] having the focus within this viewport. If no [Control] has the focus, returns null.
```
Also should mention argument somehow.

KoBeWi commented on doc/classes/Viewport.xml:
```diff
@@ -78,8 +78,8 @@
 				func _ready():
-				    await RenderingServer.frame_post_draw
-				    $Viewport.get_texture().get_image().save_png("user://Screenshot.png")
+				await RenderingServer.frame_post_draw
+				$Viewport.get_texture().get_image().save_png("user://Screenshot.png")
```
That doesn't look intended.
(and is wrong)

KoBeWi replied:
Bump .-.

redsett replied:
Gotcha. Weird, I Visual Studio refuses to let me intent those lines like that. Probably some non-sense setting. Had to open it in notepad ++ to make the change.

KoBeWi commented on scene/main/viewport.h:
```diff
@@ -355,7 +360,8 @@ class Viewport : public Node {
 		Control *mouse_click_grabber = nullptr;
 		MouseButton mouse_focus_mask = MouseButton::NONE;
-		Control *key_focus = nullptr;
+		Vector<FocusedControl> focused_controls;
```
```suggestion
		LocalVector<FocusedControl> focused_controls;
```
Can be local vector if you are not passing it anywhere, AFAIK.

KoBeWi replied:
The iterator was added, so I think this can be replaced now.

redsett replied:
This is outdated, the new version uses a HashMap.
`HashMap<int, Control *> focused_controls;`

KoBeWi commented:
Functionality-wise it's fine now. Implementation is mostly ok too, but I'm worried about the usage of loops everywhere; it might affect performance. Not sure how it could be implemented better though. *Might* be not important if you don't use many layers though.

redsett commented:
> Functionality-wise it's fine now. Implementation is mostly ok too, but I'm worried about the usage of loops everywhere; it might affect performance. Not sure how it could be implemented better though. _Might_ be not important if you don't use many layers though.

Thanks for all of the feedback! I'll get those done. And I'll also think about the for loop concern.

Calinou commented:
> Is this feature in the beta release yet? how can I test this feature?

This PR is not merged yet, so no, this feature isn't in the `master` branch or any beta.

You can download artifacts from this PR or compile it from source though: https://docs.godotengine.org/en/latest/community/contributing/testing_pull_requests.html

cybergeek1943 commented:
Ok, will this feature get into 4.0 then?

Calinou commented:
> Ok, will this feature make it into 4.0 then?

We don't know for sure yet. 4.0 is currently in feature freeze. Technically, any feature pull request that has a `4.x` milestone is postponed for a future 4.x release. In practice, some of them will still be merged for 4.0, but only on a case-by-case basis. We have to release 4.0 some day :slightly_smiling_face:

KoBeWi commented:
This PR originally had 4.0 milestone, but got postponed, because it wasn't ready to merge in time. It *might* still make it into 4.0, but it's rather significant change (even if the implementation is rather simple).

It also depends on when the author updates the PR.

redsett commented:
Applied feedback.

I didn't address the loop concern. BUT, I did think about it. `focused_controls` could easily be changed to a map<int, Control*> with some fixed allocation, then maybe add a function to set the allocation e.g. `viewport->set_focus_layer_count(int)`.

KoBeWi commented on scene/gui/control.cpp:
```diff
@@ -3054,8 +3081,10 @@ void Control::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("get_focus_layer"), &Control::get_focus_layer);
 	ClassDB::bind_method(D_METHOD("has_focus"), &Control::has_focus);
-	ClassDB::bind_method(D_METHOD("grab_focus"), &Control::grab_focus);
+	ClassDB::bind_method(D_METHOD("grab_focus", "p_auto_set_viewport_active_focus_layer"), &Control::grab_focus, DEFVAL(false));
```
```suggestion
	ClassDB::bind_method(D_METHOD("grab_focus", "auto_set_viewport_active_focus_layer"), &Control::grab_focus, DEFVAL(false));
```
`p_` prefix is only used in C++ code, this is argument name for binding.

redsett replied:
Ty!

KoBeWi commented:
> I believe focused_controls could be changed to a map<int, Control*> with some fixed allocation, then maybe add a function to set the allocation e.g. viewport->set_focus_layer_count(int). Of course this reallocation function would be slow as it would need to copy the map, reallocate, then loop through the copy and put the old values back in. OR, maybe we're fine with a dynamically sized map?

I don't really know much about allocations, but we use HashMap in many places in the engine and I never seen it as a concern. So I think it's ok.

> But there may be a few cases that may need to loop through the keys. :/

It would be best to avoid that .-.

redsett commented:
Sounds good. I'll fix that issue and see about switching to a HashMap.

redsett commented:
Updated to use a HashMap instead of  a struct. Removed/replaced all loops with map lookups.

YuriSizov commented:
Hey, thanks for your work! I don't think there is a need to update your branch at this moment. This is unlikely to be merged for 4.0, because it has a potential to create regressions and we're in a stabilization phase. We will look at this after 4.0 is out.

Keep in mind that every time you push to your branch, it runs the CI here.

KoBeWi commented:
Rebasing should only be done if there are conflicts; otherwise it's useless. And never do it using the GitHub button, as it will only cause you trouble.

While this won't likely be merged soon (as noted above), I'd like to give it another review to at least get it to approved state, as it's a nice feature. But this also requires a review from core team.

KoBeWi commented on doc/classes/Control.xml:
```diff
@@ -550,9 +550,11 @@
+			<param index="0" name="auto_set_viewport_active_focus_layer" type="bool" default="false" />
 			<description>
 				Steal the focus from another control and become the focused control (see [member focus_mode]).
+				Returns [code]true[/code] if it was able to grab focus. Would return [code]false[/code] if the control isn't set to be focused (see [member focus_mode]).
```
```suggestion
				Returns [code]true[/code] if it was able to grab focus. Returns [code]false[/code] if the control doesn't get focused (see [member focus_mode]).
```

KoBeWi commented on doc/classes/Viewport.xml:
```diff
@@ -108,16 +108,24 @@
 			<description>
-				Returns the [Control] having the focus within this viewport. If no [Control] has the focus, returns null.
+				Returns the focused [Control] within the provided focus layer in this viewport.
+				If no [Control] has the focus in the provided focus layer, returns null.
```
```suggestion
				If no [Control] has the focus in the provided focus layer, returns [code]null[/code].
```

KoBeWi commented on scene/main/viewport.cpp:
```diff
@@ -1974,14 +1976,14 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
 
-		if (gui.key_focus && !gui.key_focus->is_visible_in_tree()) {
-			gui.key_focus->release_focus();
+		if (_gui_get_active_focus_control() && !_gui_get_active_focus_control()->is_visible_in_tree()) {
```
Instead of calling the function twice, you should do something like
```C++
Control *focus_control = _gui_get_active_focus_control();
if (focus_control && ...
````

KoBeWi commented on scene/main/viewport.cpp:
```diff
@@ -1990,7 +1992,7 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
 		}
 
-		Control *from = gui.key_focus ? gui.key_focus : nullptr;
+		Control *from = _gui_get_active_focus_control() ? _gui_get_active_focus_control() : nullptr;
```
Another repeat. You can reuse the variable I mentioned above.

KoBeWi commented on scene/main/viewport.cpp:
```diff
@@ -2145,18 +2147,21 @@ void Viewport::_gui_remove_root_control(List<Control *>::Element *RI) {
 void Viewport::_gui_unfocus_control(Control *p_control) {
-	if (gui.key_focus == p_control) {
-		gui.key_focus->release_focus();
+	if (p_control && gui.focused_controls.has(p_control->get_focus_layer())) {
+		if (gui.focused_controls.get(p_control->get_focus_layer()) == p_control) {
```
Looks like it can be replaced with `_gui_control_has_focus()` (same below)

KoBeWi commented on scene/main/viewport.cpp:
```diff
@@ -2891,17 +2928,30 @@ int Viewport::gui_get_canvas_sort_index() {
+void Viewport::gui_release_focus(int p_focus_layer) {
+	HashMap<int, Control *>::Iterator F = gui.focused_controls.find(p_focus_layer);
+	if (F) {
+		if (F->value) {
```
Can `value` ever be null? The entry gets deleted when you release focus.

redsett replied:
Good point.

KoBeWi commented on doc/classes/Control.xml:
```diff
@@ -550,9 +550,11 @@
 		<method name="grab_focus">
-			<return type="void" />
+			<return type="bool" />
+			<param index="0" name="auto_set_viewport_active_focus_layer" type="bool" default="false" />
```
Might be worth documenting what is `auto_set_viewport_active_focus_layer`.

redsett commented:
Updated, applied all of your feedback @KoBeWi

Also, congratz on the 4.0 release!

KoBeWi commented on doc/classes/Control.xml:
```diff
@@ -555,10 +555,12 @@
 			<description>
-				Steal the focus from another control and become the focused control (see [member focus_mode]).
-				[b]Note:[/b] Using this method together with [method Callable.call_deferred] makes it more reliable, especially when called inside [method Node._ready].
+				Steal the focus from another control and become the focused control (see [member focus_mode]). If [param p_auto_set_viewport_active_focus_layer], then also set the focused layer to the layer this control belongs to.
```
```suggestion
				Steal the focus from another control and become the focused control (see [member focus_mode]). If [param p_auto_set_viewport_active_focus_layer] is [code]true[/code], also sets the focused layer to the layer this control belongs to.
```

redsett replied:
Bah, not sure how I missed those.

KoBeWi commented:
You missed 2 notes I left last time.

KoBeWi commented on scene/gui/control.cpp:
```diff
@@ -1907,19 +1907,43 @@ Control::FocusMode Control::get_focus_mode() const {
+	ERR_FAIL_COND_V(!is_inside_tree(), false);
 
 	if (data.focus_mode == FOCUS_NONE) {
 		WARN_PRINT("This control can't grab focus. Use set_focus_mode() to allow a control to get focus.");
```
I wonder if this warning is still necessary if the method returns bool 🤔
(up to discussion, don't change it yet)

redsett replied:
Yeah, not sure. TBH I'm not sure how a `is_inside_tree()` would fail. Maybe if you dynamically created a Control and didn't parent it to anything?

KoBeWi commented:
I was referring to these:
https://github.com/godotengine/godot/pull/62421#discussion_r981290258
https://github.com/godotengine/godot/pull/62421#discussion_r999731362
There's also this, which is worth a try:
https://github.com/godotengine/godot/pull/62421#discussion_r981340458

redsett commented:
Updated, thank you for the links! Totally missed them.

KoBeWi reviewed (approved):
Needs squashing. And likely a feedback from core team, because it modifies Viewport.

Other than that seems fine now.

redsett commented:
Did another rebase. I totally didn't comment it properly and had a ton of issues with the whole process. My git noobness certainly shows during a rebase. I think I worked out the issues though, things seem to be running as they did.

Is there any good resources for rebasing? I followed the [godot docs](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase), but it certainly doesn't go over everything. So I just followed some random tutorial.

KoBeWi commented:
Rebasing this is going to be difficult, because you used the GItHub's rebase feature and you have merge commits and other unrelated commits mixed in.

redsett commented:
Is there any decent ways to fix the rebase mishap? 🙄 Or would the best approach be to create a new fork/pull request with only the needed changes.

redsett committed:
Added back in the focus layer changes.

redsett commented:
I did a reset/force, then manually merged my changes.

redsett committed:
Merge remote-tracking branch 'upstream/master'

redsett committed:
Added error thread macros to a bunch of functions.

saki7 commented:
It seems like this change is virtually redefining the concept of focus. Are we really sure that we're good to make this the default behavior for Godot?

Personally I like the idea of allowing multiple focuses, that makes perfect sense regarding GUI controls in general. However, due to the fact that current engine assumes only one Control is focused at the same time, it can be presumed that many users have some kind of local "focused control cache" in their code which is also assuming maximum element count of 1. 

For example, one could enumerate all child controls of a node, then check `has_focus()` and store the first match to the cache. Such cases seem not trivial to me, and I think this PR might (or might not) introduce a regression on the application side.

KoBeWi commented:
This PR doesn't really change anything. It adds focus "layers", but everything is on the same layer by default, so if you don't change this property, the behavior is exactly the same as before.

YuriSizov commented:
I'm tentatively putting this on the 4.2 milestone, as I think it would be best to merge it early in a dev cycle, not late (we're close to the 4.1 release with a feature freeze this week). I would also like some input from the Input team (pun not intended but inevitable).

redsett committed:
Merge remote-tracking branch 'upstream/master'

redsett committed:
Merge remote-tracking branch 'upstream/master'

redsett committed:
Merge remote-tracking branch 'upstream/master'

Calinou commented:
@redsett Could you upload a project that can be used to test this pull request locally? Thanks in advance :slightly_smiling_face:

KoBeWi commented:
Here's a simple scene:
[FocusTest.tscn.txt](https://github.com/godotengine/godot/files/11892725/FocusTest.tscn.txt)
Use keyboard to navigate.

Sauermann commented:
Since this  PR adds quite a bit of complexity, I was exploring a different approach, namely to let the user decide, which Control-node an action-InputEvent should be routed to.
I include a MRP for testing: [KeyEventRouting.zip](https://github.com/godotengine/godot/files/12033653/KeyEventRouting.zip)

Advantage: Adds user-flexibility for deciding the target-Control-node for action-InputEvents with less complexity in the code-base
Disadvantage: Not as comfortable regarding focus handling 

``` diff
diff --git a/doc/classes/Viewport.xml b/doc/classes/Viewport.xml
index c0c53146fe..6cda9d1c0d 100644
--- a/doc/classes/Viewport.xml
+++ b/doc/classes/Viewport.xml
@@ -21,6 +21,13 @@
 		<link title="3D Viewport Scaling Demo">https://godotengine.org/asset-library/asset/586</link>
 	</tutorials>
 	<methods>
+		<method name="_route_action_event" qualifiers="virtual const">
+			<return type="Object" />
+			<param index="0" name="event" type="InputEvent" />
+			<description>
+				Virtual method to be implemented by the user. Returns the [Control]-node within the [Viewport] that the [param event] should be routed to.
+			</description>
+		</method>
 		<method name="find_world_2d" qualifiers="const">
 			<return type="World2D" />
 			<description>
diff --git a/scene/main/viewport.cpp b/scene/main/viewport.cpp
index 71bef56ebb..380780d3c4 100644
--- a/scene/main/viewport.cpp
+++ b/scene/main/viewport.cpp
@@ -2186,10 +2186,22 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
 			gui.key_focus->release_focus();
 		}
 
-		if (gui.key_focus) {
+		Control *target;
+		if (GDVIRTUAL_IS_OVERRIDDEN(_route_action_event)) {
+			Object *ret = nullptr;
+			GDVIRTUAL_CALL(_route_action_event, p_event, ret);
+			target = Object::cast_to<Control>(ret);
+			if (target && target->get_viewport() != this) {
+				target = nullptr;
+			}
+		} else {
+			target = gui.key_focus;
+		}
+
+		if (target) {
 			gui.key_event_accepted = false;
-			if (gui.key_focus->can_process()) {
-				gui.key_focus->_call_gui_input(p_event);
+			if (target->can_process()) {
+				target->_call_gui_input(p_event);
 			}
 
 			if (gui.key_event_accepted) {
@@ -4411,6 +4423,8 @@ void Viewport::_bind_methods() {
 	BIND_ENUM_CONSTANT(VRS_TEXTURE);
 	BIND_ENUM_CONSTANT(VRS_XR);
 	BIND_ENUM_CONSTANT(VRS_MAX);
+
+	GDVIRTUAL_BIND(_route_action_event, "event");
 }
 
 void Viewport::_validate_property(PropertyInfo &p_property) const {
diff --git a/scene/main/viewport.h b/scene/main/viewport.h
index e56d69a868..328afb9267 100644
--- a/scene/main/viewport.h
+++ b/scene/main/viewport.h
@@ -471,6 +471,7 @@ private:
 
 	void _process_dirty_canvas_parent_orders();
 
+	GDVIRTUAL1RC(Object *, _route_action_event, Ref<InputEvent>);
 protected:
 	void _set_size(const Size2i &p_size, const Size2i &p_size_2d_override, bool p_allocated);
 
diff --git a/scene/scene_string_names.cpp b/scene/scene_string_names.cpp
index 536ffd1fe4..c1e744ecf3 100644
--- a/scene/scene_string_names.cpp
+++ b/scene/scene_string_names.cpp
@@ -132,6 +132,7 @@ SceneStringNames::SceneStringNames() {
 
 	_unhandled_input = StaticCString::create("_unhandled_input");
 	_unhandled_key_input = StaticCString::create("_unhandled_key_input");
+	_route_action_event = StaticCString::create("_route_action_event");
 
 	changed = StaticCString::create("changed");
 	_shader_changed = StaticCString::create("_shader_changed");
diff --git a/scene/scene_string_names.h b/scene/scene_string_names.h
index ca8f7a1e7d..d83bb0cb28 100644
--- a/scene/scene_string_names.h
+++ b/scene/scene_string_names.h
@@ -126,6 +126,7 @@ public:
 	StringName _ready;
 	StringName _unhandled_input;
 	StringName _unhandled_key_input;
+	StringName _route_action_event;
 
 	StringName _pressed;
 	StringName _toggled;
```

redsett commented:
@Sauermann, does that PR allow multiple controls to be focused at once? Reading the code, it "looks" like it would jump to the another control, rather than keeping the focus on multiple controls at once. (i.e. a tab menu or something like smash bros)

Sauermann commented:
> does that PR allow multiple controls to be focused at once

No, it doesn't, graphical effects of focus would still need to be handled by the user. It is a different approach and I am not sure yet, if my approach is beneficial.

redsett committed:
Merge remote-tracking branch 'upstream/master'

redsett commented:
@Sauermann, not sure about that sample, but I can take a look.

I wasn't concerned about complexity when I did this PR, only that the frontend felt intuitive. But I totally get where you're coming from. Your suggestion/s is far less intrusive.

As this is my first PR and I'm fairly new to Godot, I'll certainly defer to your judgement. As I'm sure you've seen such things come back to bite. :)

Sauermann commented:
While it is less intrusive, it comes with other penalties, like separate Viewports, that require more rendering / CPU bandwidth.
So it isn't a simple "one is better than the other" situation.
I can see the merits in your approach.

redsett committed:
Fixed a bug where finding next and previous valid focus wasn't respecting the focus layer.

redsett commented:
@Sauermann, I f

> With the demo-project of [#62421 (comment)](https://github.com/godotengine/godot/pull/62421#issuecomment-1611108452) I get the following error, when I use the TAB-key when I am focused on the "Go Back"-button. Have not tested, if this is a problem of the demo-project or of the PR:
> 
> ```
> W 0:00:05:0143   grab_focus: This control can't grab focus. Use set_focus_mode() to allow a control to get focus.
>   <C++ Source>   scene/gui/control.cpp:2017 @ grab_focus()
> ```

This was a bug, just submitted a fix.

TheFoxKnocks commented:
This crashes a lot in the editor. I made a fresh project and created a VBoxContainer and a couple buttons. Changing the parameters of the buttons would randomly stall the editor out and have it crash. I couldn't find a common link to what causes it. Sometimes just changing custom minimum sizes breaks it, sometimes just tabbing to a new button breaks it, sometimes these kinds of things work fine.

The most volatile was changing the focus layer parameter which seemed to crash it more frequently, but curiously not on all buttons. Some buttons I could change this freely, other buttons changing the parameter at all would crash 100% of the time.

Literally just a VBox container and a few buttons, nothing else.

redsett committed:
Merge remote-tracking branch 'upstream/master'

redsett commented:
Hmmm. Just tested as well. Clearly something has gone wrong. I'll take a look.

redsett committed:
Merge branch 'master' of https://github.com/godotengine/godot

redsett committed:
Fixed a bug when releasing focus would loop into itself in certain situations.

The issue was that the control that is trying to be released was not getting removed before it would try to redraw. So the redraw would eventually call it to get released. Over and over.

redsett commented:
@TheFoxKnocks, I found and fixed that bug. The issue was that the control that is trying to be released was not getting removed before it would try to redraw. So the redraw would eventually call it to get released again. Over and over.

Also, thank you for testing!

TheFoxKnocks commented:
> @TheFoxKnocks, I found and fixed that bug. The issue was that the control that is trying to be released was not getting removed before it would try to redraw. So the redraw would eventually call it to get released again. Over and over.
> 
> Also, thank you for testing!

I actually very much need what you're creating for my own project, so I have a vested interest in seeing it do well. Thank you for both creating it and fixing bugs in such a timely manner. I'll look more into it soon.

redsett commented:
> > @TheFoxKnocks, I found and fixed that bug. The issue was that the control that is trying to be released was not getting removed before it would try to redraw. So the redraw would eventually call it to get released again. Over and over.
> > Also, thank you for testing!
> 
> I actually very much need what you're creating for my own project, so I have a vested interest in seeing it do well. Thank you for both creating it and fixing bugs in such a timely manner. I'll look more into it soon.
> 
> EDIT: Damn, I just noticed that this doesn't support simultaneous input, and the other implementation linked to is abandoned since 2019.

By "simultaneous input", you are referring to multiple users navigating the same buttons? Yeah, this wouldn't support that.

If you're referring to co-op type menu where each user has their own menu. Then you could set a different focus_layer on the buttons of each menu. Then when an input happens, you could set_active_focus_layer() to whoever pressed the input.

Sauermann commented on scene/main/viewport.h:
```diff
@@ -591,8 +595,11 @@ class Viewport : public Node {
+	int gui_get_active_focus_layer() const;
+	void gui_set_active_focus_layer(int p_focus_layer);
+
+	void gui_release_focus(int p_focus_layer);
```
Have you thought about providing a default-value `gui_release_focus(int p_focus_layer = 0);`?

redsett replied:
No reason not to. Will add.

Sauermann reviewed (commented):
I had a look at the code and it looks like the PR works as advertised. Also the editor seems to work without problems from a few quick tests.
From what I can tell, the behavior doesn't change, if the newly introduced functionality (focus layers) isn't used, so it stays compatible to current behavior. However it changes the API of two functions that are exposed to GDScript.

Since #79480 tries to accomplish the same goal in a different way, I want to include a comparison of the two PRs. Both have advantages and disadvantages, so it is difficult for me to tell, which of the two approaches would be better suited.

| Topic | This PR | #79480 |
| --- | --- | --- |
| Size of change | +176 −64 | +17 −10 |
| Behavior change | no, current behavior stays the same, only adds functionality | method for focusing Controls is redefined and becomes more complex |
| Chance of introducing regressions | low | medium to high |
| API change | `gui_get_focus_owner` and `gui_release_focus` change their GDScript API signature | no API change |
| Maintenance | the concept of focus layers adds complexity to Viewports and Controls | no additional concepts are introduced |
| Performance | should have no to little impact | Every group of linked Controls requires a separate SubViewport. Additional Viewports have a negative impact on rendering performance |
| Screen space real estate | Adds focus_layer to Control inspector | no changes to inspector |
| Discoverability | focus layer is introduced as new variable in the inspector, so it can be discovered easily | Can only be discovered when multiple SubViewports are used, so the functionality can be considered difficult to discover |

redsett commented on scene/gui/control.cpp:
```diff
@@ -1985,21 +1985,45 @@ Control::FocusMode Control::get_focus_mode() const {
+		if (next && next != this) {
+			next->grab_focus(false);
+		} else {
+			release_focus();
```
That's fair. I was thinking in terms of a game, where a button is always in focus. Or at least a game that supports controllers.

Alternatively this behavior could be gated behind a bool. Or just let the user do this logic themselves if that's what they want.

But I agree that forcing this logic is probably not the correct thing.

redsett replied:
I removed that logic. It now just releases focus if it has focus.

redsett commented on scene/main/viewport.cpp:
```diff
@@ -4189,13 +4236,16 @@ void Viewport::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("gui_set_active_focus_layer", "focus_layer"), &Viewport::gui_set_active_focus_layer);
+
+	ClassDB::bind_method(D_METHOD("gui_release_focus", "focus_layer"), &Viewport::gui_release_focus);
+	ClassDB::bind_method(D_METHOD("gui_get_focus_owner", "focus_layer"), &Viewport::gui_get_focus_owner);
```
Because it didn't know that was an option. :)

I see an example in control.cpp. Will add here as suggested.

redsett commented on scene/main/viewport.cpp:
```diff
@@ -2428,24 +2432,51 @@ Window *Viewport::get_base_window() const {
-	return gui.key_focus == p_control;
+bool Viewport::_gui_has_active_focus_layer() const {
+	return gui.focused_controls.has(gui_get_active_focus_layer());
+}
```
Good catch. At some point I probably had a use for it. But clearly not now.

redsett committed:
Applied code review feedback from Sauermann.

gui_release_focus now defaults to focus layer 0.
Removed _gui_has_active_focus_layer because it was never used.
Added default values to the bind methods for release and get focus owner.
Removed the logic in set_focus_layer that would find the next valid control before releasing focus.

redsett commented:
Wow, nice comparison Sauermann.

djpeach commented:
I can imagine a world where both PRs are merged. Not sure on the implementation effects on each other, but as far as usability from the standpoint of a godot game dev, i can picture using multiple viewports as in #79480 with #78762 for local split screen, or in general for UI’s with a bigger differentiation between each other, and then using this PR’s control focus layers for related nested UI’s like in the description. 

Wonder if it is possible to have it all, and allow the user to pick. @Sauermann ’s comparison above would be a great starting point to adding a description of the the tradeoffs between the two cases to the docs for users to read and make their own decisions based on their needs

djpeach commented:
After playing around with Sauermann's implementation (split into #78762 and #79480) allows nested input layers/screens without losing focus, local splitscreen menu's, and really whatever the dev wants to do. 

[Demo](https://github.com/godotengine/godot/pull/79480#issuecomment-1663081085)

redsett commented:
> After playing around with Sauermann's implementation (split into #78762 and #79480) allows nested input layers/screens without losing focus, local splitscreen menu's, and really whatever the dev wants to do.
> 
> [Demo](https://github.com/godotengine/godot/pull/79480#issuecomment-1663081085)

Yeah, both CL seem to achieve the same thing in different ways.

You can do multiplayer type stuff with this CL as well. You just have to set_active_focus_layer() when an input happens based on which player pressed the input.

I.e
get_viewport().set_active_focus_layer(player_input_index)

redsett commented:
@djpeach, here's your example using this PR.

![image](https://github.com/godotengine/godot/assets/17795893/01ebef30-24b2-4ffd-b546-18081163a584)
[LocalSplitScreenMenu_UsingFocusLayers.zip](https://github.com/godotengine/godot/files/12262602/LocalSplitScreenMenu_UsingFocusLayers.zip)

Technically the sub viewports are not actually needed in this example.

djpeach commented:
That's awesome. I was wondering how this PR would play with the others. I can picture using event filtering for the sub-viewports and then the focus layers for nested UI's for each player

reduz commented:
I think there is a lot of confusion in this pull request. So I will give my view on it.

1. What this PR solves is very easy to workaround. You don't need focus to highlight something selected. Plenty of controls to do this already without the need of this feature.
2. You also don't need layers to restraint focus movement to an area, you can override this with next/prev focus. If not enough, I think a separate feature to tell a control to restrict focus movement to children would be more useful and more explicit.
3. What pretty much everyone here wants by checking the discussion is having independent focuses per viewport for split screen. **This is not what this PR is meant for**. I think a separate PR should be opened to address this more specifically, doing the following:
  A. The ability to assign a layer mask to events in the input event editor.
  B. The ability to assign a layer mask to viewports as a viewport property, so events are filtered.
  C. The ability to set a Viewport to manage its own focus, that is, its focus can't be stolen by other viewports and other viewports can't steal its focus.

As such I think this PR is not necessary. My view is that there is mostly misunderstanding around it and that it should be its own separate proposal first to discuss this feature specifically, leaving aside the separate focus per viewport, which is not related.

KoBeWi commented:
> You also don't need layers to restraint focus movement to an area, you can override this with next/prev focus.

This does not solve the problem. It's impossible to restrict the focus movement with the current system.

> If not enough, I think a separate feature to tell a control to restrict focus movement to children would be more useful and more explicit.

Except it would force you to assign all focus neighbors explicitly, which is not very feasible in complex menus, especially when you want to later rearrange stuff or the menu itself is dynamic.

> What pretty much everyone here wants by checking the discussion is having independent focuses per viewport for split screen.

tbh I did not even consider that when doing my review. I myself ran into the problem of not being able to restrict focus to specific areas, which I think is a common issue if you rely on the built-in Control system for menu layouts, and this PR solves that.

Sauermann commented:
> > If not enough, I think a separate feature to tell a control to restrict focus movement to children would be more useful and more explicit.
> 
> Except it would force you to assign all focus neighbors explicitly, which is not very feasible in complex menus, especially when you want to later rearrange stuff or the menu itself is dynamic.

I believe, that I interpret the idea differently than you, KoBeWi:
When a focus movement happens and a parent Control `P` has a new flag `restrict_focus_movement_to_children` set to true, then the focus movement-algorithm (`Control::find_next_valid_focus`, `Control::_get_focus_neighbor`, ...) should consider as valid Control-nodes only children of the node `P`.

@reduz 

> I think a separate PR should be opened to address this more specifically, doing the following:
> A. The ability to assign a layer mask to events in the input event editor.
> B. The ability to assign a layer mask to viewports as a viewport property, so events are filtered.

The Input Event editor allows configuring only `InputEventAction`. I wonder if it wouldn't be better to allow SubViewports to filter all `InputEvents`. #78762 implements this idea.

> C. The ability to set a Viewport to manage its own focus, that is, its focus can't be stolen by other viewports and that viewports can't steal others focus.

I believe that it is feasible to remove focus-stealing completely. See #79480 for an implementation of that.

KoBeWi commented:
> When a focus movement happens and a parent Control P has a new flag restrict_focus_movement_to_children set to true, then the focus movement-algorithm (Control::find_next_valid_focus, Control::_get_focus_neighbor, ...) should consider as valid Control-nodes only children of the node P.

Right, I didn't notice the "children" part. It would need to be "descendants", as the node you want focused doesn't need to be a direct child (not uncommon when you put too many containers for styling), but it sounds acceptable.

redsett commented:
I'm personally fine with the sub viewport method that Saurmann made, [79480](https://github.com/godotengine/godot/pull/79480).

You could do a normal settings menu or smash bros menu with the sub viewport method. A bit unintuitive, but flexible.

TheFoxKnocks commented:
I'm a little confused on which implementation, if any, is looking to be official. I was looking forward to a 4.2 implementation of proper local coop input support, but it looks like that might not be happening now? I'm not fully sure.

Don't get me wrong, not rushing or being ungrateful, really just trying to figure out where this stands.

redsett commented:
> I'm a little confused on which implementation, if any, is looking to be official. I was looking forward to a 4.2 implementation of proper local coop input support, but it looks like that might not be happening now? I'm not fully sure.
> 
> Don't get me wrong, not rushing or being ungrateful, really just trying to figure out where this stands.

It sounds like this PR adds too much complexity. But with these PR that Sauermann made, you can achieve the same result for local/and or split screen co-op. [78762](https://github.com/godotengine/godot/pull/78762) [79480](https://github.com/godotengine/godot/pull/79480)

You could just put each player's UI on a separate subviewport, then pass the input for each player to the correct viewport.

juliohq commented:
I like the idea of having different focus layers. I need this PR to be merged for a project I'm working on, which doesn't have a complex UI but needs tabs that should be focused at the same time as other controls.

I tried to implement them by hand but didn't have much luck, since trying to split the logic is overly complex and still doesn't provide me with useful features the built-in focus provides, like being able to use the built-in `ui_*` input actions on them and hence requiring more manual work.

Chaosus commented:
@redsett This PR need to be rebased, just a remainder

redsett commented:
> @redsett This PR need to be rebased, just a reminder

Thanks. I'll rebase if you guys decide to go this route. But it sounds like they want to go with another solution.

TheFoxKnocks commented:
Am I correct in assuming that ultimately neither choice was decided to be the right one and Godot still has no way to handle multiple focus targets?

Sauermann commented:
A discussion on which solution would be preferable is still pending.
So the current status is, that multiple focus targets are not supported.

Mickeon commented on doc/classes/Control.xml:
```diff
@@ -565,10 +565,12 @@
 			<description>
-				Steal the focus from another control and become the focused control (see [member focus_mode]).
-				[b]Note:[/b] Using this method together with [method Callable.call_deferred] makes it more reliable, especially when called inside [method Node._ready].
+				Steal the focus from another control and become the focused control (see [member focus_mode]). If [param auto_set_viewport_active_focus_layer] is [code]true[/code], also sets the focused layer to the layer this control belongs to.
+				Returns [code]true[/code] if it was able to grab focus. Returns [code]false[/code] if the control doesn't get focused (see [member focus_mode]).
+				[b]Note[/b]: Using this method together with [method Callable.call_deferred] makes it more reliable, especially when called inside [method Node._ready].
```
This is a lovely amount of detail and it's a shame it's locked behind this contested PR. Should be a separate PR outright.
